### PR TITLE
Integrate trn generation api

### PR DESCRIPTION
**Pull request by hortha**
### Context

Generate TRN from a REST API to be able to assign it immediately when creating a new teacher in DQT
rather than the Create / Update overhead which happens at the moment

### Changes proposed in this pull request

A feature management controlled feature to be able to generate a TRN by calling the TRN Generation API.

### Guidance to review

To use the TRN Generation API to generete a TRN requires an appropriate deployment for that API + the feature to be enabled in DQT API.

### Checklist

-   [ x] Attach to Trello card
-   [ x] Rebased master
-   [x ] Cleaned commit history
-   [ x] Tested by running locally

### CODE REVIEW

**Comment by hortha** [comment]
[Spike to integrate TRN Generation API into DQT API](https://trello.com/c/cVgMq4Ja/394-spike-to-integrate-trn-generation-api-into-dqt-api)

File: DqtApi/src/DqtApi/DataStore/Crm/DataverseAdapter.CreateTeacher.cs [hunk 1]
```diff
@@ -38,14 +38,23 @@ public async Task<CreateTeacherResult> CreateTeacher(CreateTeacherCommand comman
                 return (CreateTeacherResult.Failed(failedReasons), null);
             }
 
+            var newContact = helper.CreateContactEntity();
+            string trn = null;
+            bool isUseTrnGenerationApiFeatureEnabled = await _featureManager.IsEnabledAsync(FeatureFlags.UseTrnGenerationApi);
+            if (isUseTrnGenerationApiFeatureEnabled)
+            {
+                trn = await _trnGenerationApiClient.GenerateTrnAsync();
```
Comment (gunndabad) [inline DqtApi/src/DqtApi/DataStore/Crm/DataverseAdapter.CreateTeacher.cs:46]: This needs to move further down and only be called when `allocateTrn` is `false` (i.e. we haven't found a duplicate).

File: DqtApi/src/DqtApi/DataStore/Crm/DataverseAdapter.CreateTeacher.cs [hunk 2]
```diff
@@ -38,14 +38,23 @@ public async Task<CreateTeacherResult> CreateTeacher(CreateTeacherCommand comman
                 return (CreateTeacherResult.Failed(failedReasons), null);
             }
 
+            var newContact = helper.CreateContactEntity();
+            string trn = null;
+            bool isUseTrnGenerationApiFeatureEnabled = await _featureManager.IsEnabledAsync(FeatureFlags.UseTrnGenerationApi);
```
Comment (gunndabad) [inline DqtApi/src/DqtApi/DataStore/Crm/DataverseAdapter.CreateTeacher.cs:43]:
```suggestion
            bool useTrnGenerationApi = await _featureManager.IsEnabledAsync(FeatureFlags.UseTrnGenerationApi);
```

File: DqtApi/src/DqtApi/Services/TrnGenerationApi/ITrnGenerationApiClient.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿namespace DqtApi.Services.TrnGenerationApi
+{
+    using System.Threading.Tasks;
+
+    public interface ITrnGenerationApiClient
+    {
+        Task<string> GenerateTrnAsync();
```
Comment (gunndabad) [inline DqtApi/src/DqtApi/Services/TrnGenerationApi/ITrnGenerationApiClient.cs:7]: We don't tend to add an `Async` suffix to method names
```suggestion
        Task<string> GenerateTrn();
```

File: DqtApi/src/DqtApi/Services/TrnGenerationApi/ITrnGenerationApiClient.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿namespace DqtApi.Services.TrnGenerationApi
+{
+    using System.Threading.Tasks;
```
Comment (gunndabad) [inline DqtApi/src/DqtApi/Services/TrnGenerationApi/ITrnGenerationApiClient.cs:3]: Nit: `using`s outside namespace

File: DqtApi/src/DqtApi/DataStore/Crm/DataverseAdapter.CreateTeacher.cs (hunk 2 above)
Comment (hortha) [reply DqtApi/src/DqtApi/DataStore/Crm/DataverseAdapter.CreateTeacher.cs]: Fixed

File: DqtApi/src/DqtApi/Services/TrnGenerationApi/ITrnGenerationApiClient.cs (hunk 3 above)
Comment (hortha) [reply DqtApi/src/DqtApi/Services/TrnGenerationApi/ITrnGenerationApiClient.cs]: Amended

File: DqtApi/src/DqtApi/Services/TrnGenerationApi/ITrnGenerationApiClient.cs (hunk 4 above)
Comment (hortha) [reply DqtApi/src/DqtApi/Services/TrnGenerationApi/ITrnGenerationApiClient.cs]: Fixed and sorted my Visuals Studio settings!

File: DqtApi/src/DqtApi/DataStore/Crm/DataverseAdapter.CreateTeacher.cs (hunk 1 above)
Comment (hortha) [reply DqtApi/src/DqtApi/DataStore/Crm/DataverseAdapter.CreateTeacher.cs]: Moved things around

**Review by gunndabad — APPROVED** [review]

**Pull request merged**